drivers/shared/procio/io.go

```diff
@@ -0,0 +1,165 @@
+package procio
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/hashicorp/nomad/client/lib/fifo"
+)
+
+var bufPool = sync.Pool{
+	New: func() interface{} {
+		buffer := make([]byte, 32<<10)
+		return &buffer
+	},
+}
+
+type SetWritersCB func(out, err io.WriteCloser)
+
+// IO is an interface around process IO. Currently this includes stdout and
+// stderr.
+type IO interface {
+	io.Closer
+	Stdout() io.ReadCloser
+	Stderr() io.ReadCloser
+	Set(SetWritersCB)
+}
+
+type IOType string
+
+const (
+	IOTypeEmpty  IOType = ""
+	IOTypeFIFO   IOType = "fifo"
+	IOTypeBuffer IOType = "buffer"
+)
+
+// Stdio is the configuration for where to write standard process output to
+type Stdio struct {
+	IOType IOType
+	Stdout string
+	Stderr string
+}
+
+// IsNull returns true if the stdio is not defined
+func (s Stdio) IsNull() bool {
+	return s.IOType == "" && s.Stdout == "" && s.Stderr == ""
+}
+
+// ProcessIO wraps an IO interface and and sends output to the configured
+// destination
+type ProcessIO struct {
+	io IO
+
+	copyLoop bool
+	stdio    Stdio
```

Comment:
`IO` is overused in this package, making it hard to follow the distinctions, e.g. `io` vs `stdio`.  Maybe we can name them by usage? `internalIO` vs `configIO`?  Or a more descriptive comment?

Reply:
`sink` and `source` may be good variable names to illustrate which direction the data flows.